clients/src/main/java/org/apache/kafka/common/network/Selector.java

```diff
@@ -292,15 +292,25 @@ public void close() {
      */
     public void send(Send send) {
         String connectionId = send.destination();
-        if (closingChannels.containsKey(connectionId))
+        KafkaChannel channel = channelOrFail(connectionId);
+        if (closingChannels.containsKey(connectionId)) {
+            channel.state(ChannelState.FAILED_SEND);
+            // ensure notification via `disconnected`
             this.failedSends.add(connectionId);
-        else {
-            KafkaChannel channel = channelOrFail(connectionId, false);
+        } else {
             try {
                 channel.setSend(send);
-            } catch (CancelledKeyException e) {
+            } catch (Exception e) {
+                // update the state for consistency, the channel will be discarded after `close`
+                channel.state(ChannelState.FAILED_SEND);
+                // ensure notification via `disconnected`
                 this.failedSends.add(connectionId);
                 close(channel, false);
+                if (!(e instanceof CancelledKeyException)) {
+                    log.error("Unexpected exception during send, closing connection {} and rethrowing exception {}",
+                            connectionId, e);
+                    throw e;
```

Comment:
Hmm, if there is any exception from here, we probably want to bubble it up to the caller. For example, in SocketServer.processNewResponses(), if send() hits any exception, we want to call updateRequestMetrics(() and avoid updating inflightResponses, and move on to the next response in the queue in the same loop.

Reply:
@junrao since the connection id will appear in `disconnected`, the caller will update `inflightResponses` and `updateRequestMetrics`. So perhaps it's a bit confusing for the exception to be bubbled up _and_ for the connection id to appear in disconnected. Maybe we should pick one. Since the caller can't really do much besides logging, we should just log and add to disconnected.